index.bs

```diff
@@ -535,167 +738,319 @@ To determine whether an |attribute| <dfn>contains a javascript: URL</dfn>:
 
-<div algorithm>
-To <dfn for="SanitizerConfig">allow an element</dfn> |element| with a {{SanitizerConfig}} |configuration|, do:
+The configuration modifier methods are methods on {{Sanitizer}} that modify its configuration.
+They return a boolean which informs the caller whether the configuration was modified or not.
 
-1. Set |element| to the result of [=canonicalize a sanitizer element with attributes=] with |element|.
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/elements}}"].
-1. [=list/Append=] |element| to |configuration|["{{SanitizerConfig/elements}}"].
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/removeElements}}"].
-1. [=SanitizerConfig/Remove=] |element| from |configuration|["{{SanitizerConfig/replaceWithChildrenElements}}"].
+<div class=example>
+* `new Sanitizer({elements: ["div"]}).allowElement("p");` &mdash; Returns true. Allows "div" and "p".
+* `new Sanitizer({elements: ["div"]}).removeElement("p");` &mdash; Returns false. Allows "div".
 
-NOTE: Handling of [=allowElement=] is a little more complicated than the other
-    methods, because the element allow list can have per-element allow- and
-    remove-attribute lists. We first remove the given element from the list
-    before then adding it, which has the effect of re-setting (rather than
-    merging or elsehow modifying) the per-element list to whatever is passed
-    in. In other words, the per-element allow- and remove-lists can only be
-    set as a whole.
+</div>
+
+<div algorithm>
+To <dfn for="SanitizerConfig">allow an element</dfn> |element| with a
+{{SanitizerConfig}} |configuration|, do:
+
+<div class=note><span class=marker>Note:</span>
+This requires that we distinguish 4 cases:
+* Whether we have a global allow- or remove-list, and
+* whether they already contain |element| or not.
+
+The case where an |element| with per-element allow- or remove-lists is added to a configuration
+with a global element allow-list requires additional work, as we need to ensure those do not
+violate our validity criteria. To do so, we will remove elements from the per-element allow- or
+remove list that conflict with the existing global lists.
+</div>
 
-NOTE: [=SanitizerConfig/Remove=] matches on name and namespace, so adding an
-    element with attributes would still remove the matching element from the
-    {{SanitizerConfig/removeElements}} and {{SanitizerConfig/replaceWithChildrenElements}} lists.
+1. Set |element| to the result of [=canonicalize a sanitizer element with attributes=] with
+    |element|.
+1. [=SanitizerConfig/Remove=] |element| from
+    |configuration|["{{SanitizerConfig/replaceWithChildrenElements}}"].
+1. If |configuration|["{{SanitizerConfig/elements}}"] [=map/exists=]:
+    1. If |configuration|["{{SanitizerConfig/elements}}"] does not [=map/contain=] |element|:
+        1. [=Comment=]: This is the case with a global allow-list that does not yet contain
```

Comment:
This seems wrong to me. We always have to make sure that the attributes are valid.